In [1]:
import pandas as pd

pd.set_option('display.max_columns', 500)

In [2]:
data = pd.read_csv("data/processed_data_2022-06-13.csv")
data.head()

sofifa_id                                         player_url  short_name  \
0     167575  https://sofifa.com/player/167575/anton-shunin/...   A. Shunin   
1     205211  https://sofifa.com/player/205211/milan-borjan/...   M. Borjan   
2     244700  https://sofifa.com/player/244700/yuriy-dyupin/...   Y. Dyupin   
3     199904  https://sofifa.com/player/199904/jesse-joronen...  J. Joronen   
4     240225  https://sofifa.com/player/240225/matvey-safono...  M. Safonov   

             long_name  overall  potential  value_eur  wage_eur         dob  \
0         Anton Shunin       76         76        NaN       NaN  1987-01-27   
1         Milan Borjan       76         76        NaN       NaN  1987-10-23   
2         Yuriy Dyupin       74         74        NaN       NaN  1988-03-17   
3  Jesse Pekka Joronen       72         72        NaN       NaN  1993-03-21   
4       Matvey Safonov       72         82        NaN       NaN  1999-02-25   

   weight_kg  height_cm  club_team_id club_position  club_jersey_number  \
0         88        191           NaN           NaN                 NaN   
1         84        196           NaN           NaN                 NaN   
2         84        187           NaN           NaN                 NaN   
3         89        197           NaN           NaN                 NaN   
4         82        192           NaN           NaN                 NaN   

   preferred_foot  weak_foot  skill_moves  international_reputation  \
0               0          3            1                         1   
1               1          2            1                         1   
2               1          3            1                         1   
3               1          3            1                         1   
4               1          2            1                         1   

   work_rate  body_type  release_clause_eur  pace  shooting  passing  \
0          0          0                 NaN   NaN       NaN      NaN   
1          0          0                 NaN   NaN       NaN      NaN   
2          0          0                 NaN   NaN       NaN      NaN   
3          0          0                 NaN   NaN       NaN      NaN   
4          0          0                 NaN   NaN       NaN      NaN   

   dribbling  defending  physic  attacking_crossing  attacking_finishing  \
0        NaN        NaN     NaN                  18                   11   
1        NaN        NaN     NaN                  15                   12   
2        NaN        NaN     NaN                  12                    6   
3        NaN        NaN     NaN                  15                   13   
4        NaN        NaN     NaN                  13                    7   

   attacking_heading_accuracy  skill_dribbling  skill_curve  \
0                          12               14           15   
1                          24               28           16   
2                          10                7           14   
3                          18               16           14   
4                          14               10           11   

   skill_fk_accuracy  skill_long_passing  skill_ball_control  \
0                 13                  30                  20   
1                 16                  48                  23   
2                 11                  22                  12   
3                 12                  25                  16   
4                 12                  26                  18   

   movement_acceleration  movement_sprint_speed  movement_agility  \
0                     50                     39                43   
1                     47                     40                40   
2                     20                     21                27   
3                     48                     45                50   
4                     19                     22                33   

   movement_reactions  movement_balance  power_shot_power  power_jumping  \
0                  71         

In [3]:
def count_na(df: pd.DataFrame) -> None:
    nan_sum = df.isna().sum()
    display(nan_sum[nan_sum > 0])

count_na(data)

value_eur                74
wage_eur                 61
club_team_id             61
club_position            61
club_jersey_number       61
release_clause_eur     1176
pace                   2132
shooting               2132
passing                2132
dribbling              2132
defending              2132
physic                 2132
goalkeeping_speed     17107
dtype: int64

Analisando essa tabela com os valores nulos podemos perceber que alguns jogadores do dataset não possuem um time atual, isso explica
`wage_eur`, `club_team_id`, `club_position`, `club_jersey_number` terem a mesma quantidade de valores nulos (61),
porém `value_eur` deveria seguir esse mesmo padrão, assim podemos dizer que 13 (74 - 61) jogadores que estão ativos não possuem um valor,
portanto precisamos prever esses dados.

Também podemos perceber que temos 2132 jogadores não possuem `pace`, `shooting`, `passing`, `dribbling`, `defending`, `physic`. Vamos analisar esses jogadores

In [4]:
missing_main_attr_data = data[data.pace.isna()]

n_rows = missing_main_attr_data.shape[0]
n_gk = missing_main_attr_data.GK.sum()
print(f"porcentagem de goleiros {n_gk/n_rows * 100}%")
missing_main_attr_data

porcentagem de goleiros 100.0%


sofifa_id                                         player_url  \
0        167575  https://sofifa.com/player/167575/anton-shunin/...   
1        205211  https://sofifa.com/player/205211/milan-borjan/...   
2        244700  https://sofifa.com/player/244700/yuriy-dyupin/...   
3        199904  https://sofifa.com/player/199904/jesse-joronen...   
4        240225  https://sofifa.com/player/240225/matvey-safono...   
...         ...                                                ...   
2127     257618  https://sofifa.com/player/257618/armin-gremsl/...   
2128     262424  https://sofifa.com/player/262424/sorin-mogosan...   
2129     262420  https://sofifa.com/player/262420/robert-popa/2...   
2130     210022  https://sofifa.com/player/210022/philip-sprint...   
2131     244417  https://sofifa.com/player/244417/julian-krahl/...   

       short_name            long_name  overall  potential  value_eur  \
0       A. Shunin         Anton Shunin       76         76        NaN   
1       M. Borjan         Milan Borjan       76         76        NaN   
2       Y. Dyupin         Yuriy Dyupin       74         74        NaN   
3      J. Joronen  Jesse Pekka Joronen       72         72        NaN   
4      M. Safonov       Matvey Safonov       72         82        NaN   
...           ...                  ...      ...        ...        ...   
2127    A. Gremsl         Armin Gremsl       63         66   475000.0   
2128  S. Mogoșanu       Sorin Mogoșanu       58         60   170000.0   
2129      R. Popa          Robert Popa       56         75   325000.0   
2130    P. Sprint        Philip Sprint       62         64   350000.0   
2131     J. Krahl         Julian Krahl       62         76   850000.0   

      wage_eur         dob  weight_kg  height_cm  club_team_id club_position  \
0          NaN  1987-01-27         88        191           NaN           NaN   
1          NaN  1987-10-23         84        196           NaN           NaN   
2          NaN  1988-03-17         84        187           NaN           NaN   
3          NaN  1993-03-21         89        197           NaN           NaN   
4          NaN  1999-02-25         82        192           NaN           NaN   
...        ...         ...        ...        ...           ...           ...   
2127    2000.0  1994-08-13         85        190      115716.0           RES   
2128    1000.0  1992-11-22         77        180      115716.0            GK   
2129     500.0  2003-03-05         83        190      115716.0           SUB   
2130     700.0  1993-06-27         98        196      115820.0            GK   
2131     500.0  2000-01-22         90        194      115820.0           SUB   

      club_jersey_number  preferred_foot  weak_foot  skill_moves  \
0                    NaN               0          3            1   
1                    NaN               1          2            1   
2                    NaN               1          3            1   
3                    NaN               1          3            1   
4                    NaN               1          2            1   
...                  ...             ...        ...          ...   
2127                94.0               0          2            1   
2128                 1.0               1          1            1   
2129                31.0               1          2            1   
2130                 1.0               1          2            1   
2131                13.0               1          3            1   

      international_reputation  work_rate  body_type  release_clause_eur  \
0                            1          0          0                 NaN   
1                            1          0          0                 NaN   
2                            1          0          0                 NaN   
3                            1          0          0                 NaN   
4                            1          0          0                 NaN   
...                        ...        ...        ...  

Como vimos que os goleiros não tem os atributos listados acima, mas possuem um overall em cada posição, seria interessante prever esses atributos.

In [5]:
missing_gk_speed_data = data[data.goalkeeping_speed.isna()]
sum_of_missing_data = missing_gk_speed_data.shape[0] + missing_main_attr_data.shape[0]
print(f"Jogadores sem gk speed + jogadores sem atributos primários: {sum_of_missing_data}")
print(f"Numero de jogadores: {data.shape[0]}")

Jogadores sem gk speed + jogadores sem atributos primários: 19239
Numero de jogadores: 19239


Assim podemos ver que todos os jogadores que não são goleiros não possuem `goalkeeping_speed`

## Prevendo valor dos jogadores
Nessa etapa vamos criar um modelo para prever o valor dos 13 jogadores que não possuem valor

In [6]:
players_with_club_and_no_value = data.query("not club_team_id.isna() and value_eur.isna()").copy()
players_with_club_and_no_value

sofifa_id                                         player_url  \
11929     115909  https://sofifa.com/player/115909/ruben-castro-...   
11987     124344  https://sofifa.com/player/124344/ibrahim-oztur...   
12264     153066  https://sofifa.com/player/153066/lucas-licht/2...   
13149     152912  https://sofifa.com/player/152912/jose-sand/220002   
13177     110381  https://sofifa.com/player/110381/maximiliano-r...   
14103     262371  https://sofifa.com/player/262371/craig-dundas/...   
14611     140222  https://sofifa.com/player/140222/paulo-da-silv...   
15125     138703  https://sofifa.com/player/138703/andres-daless...   
16065     158293  https://sofifa.com/player/158293/zhi-zheng/220002   
16498       2702  https://sofifa.com/player/2702/kevin-ellison/2...   
16814      29118  https://sofifa.com/player/29118/shinji-ono/220002   
18159      29552  https://sofifa.com/player/29552/shunsuke-nakam...   
18166     254704  https://sofifa.com/player/254704/kazuyoshi-miu...   

            short_name                     long_name  overall  potential  \
11929     Rubén Castro           Rubén Castro Martín       69         69   
11987        İ. Öztürk                İbrahim Öztürk       67         67   
12264         L. Licht            Lucas Matías Licht       69         69   
13149          J. Sand             José Gustavo Sand       75         75   
13177     M. Rodríguez   Maximiliano Rubén Rodríguez       73         73   
14103        C. Dundas                  Craig Dundas       56         56   
14611      P. Da Silva  Paulo César Da Silva Barrios       67         67   
15125  A. D'Alessandro   Andrés Nicolás D'Alessandro       75         75   
16065        Zheng Zhi                            郑智       67         67   
16498       K. Ellison           Kevin Keith Ellison       58         58   
16814           S. Ono                         小野 伸二       58         58   
18159      S. Nakamura                         中村 俊輔       63         63   
18166         K. Miura                         三浦 知良       59         59   

       value_eur  wage_eur         dob  weight_kg  height_cm  club_team_id  \
11929        NaN    3000.0  1981-06-27         68        169      100851.0   
11987        NaN    3000.0  1981-06-21         80        186      101006.0   
12264        NaN    5000.0  1981-04-06         72        174      101084.0   
13149        NaN   13000.0  1980-07-17         79        182      110395.0   
13177        NaN   10000.0  1981-01-02         74        173      110396.0   
14103        NaN    1000.0  1981-02-16         89        188      110799.0   
14611        NaN     500.0  1980-02-01         76        180      111007.0   
15125        NaN     900.0  1981-04-15         69        174      111325.0   
16065        NaN    5000.0  1980-08-20         75        180      111839.0   
16498        NaN    2000.0  1979-02-23         76        185      112254.0   
16814        NaN     800.0  1979-09-27         76        175      112444.0   
18159        NaN     800.0  1978-06-24         71        178      113197.0   
18166        NaN     700.0  1967-02-26         72        177      113197.0   

      club_position  club_jersey_number  preferred_foot  weak_foot  \
11929            ST                 7.0               1          3   
11987            CB                38.0               1          2   
12264           SUB                25.0               0          4   
13149            LS                 9.0               1          4   
13177           SUB                11.0               1          4   
14103           RES                14.0               1          3   
14611           LCB                 3.0               1          3   
15125           SUB                10.0               0          3   
16065           LCM                10.0               1          4   
16498           SUB                22.0               0          3   
16814           SUB                44.0               1          4   
18159         

In [7]:
data_without_players_with_club_and_no_value = data.drop(players_with_club_and_no_value.index)
data_to_model = data_without_players_with_club_and_no_value.query("not value_eur.isna()").copy()
data_to_model.drop(columns=["sofifa_id", "player_url", "short_name", "long_name", "dob", "club_position"], inplace=True)
data_to_model

overall  potential   value_eur  wage_eur  weight_kg  height_cm  \
8           83         84  28000000.0   87000.0         83        190   
9           74         79   5000000.0   36000.0         77        188   
10          61         77    750000.0    3000.0         83        196   
11          84         85  33500000.0   81000.0         88        195   
12          72         73   1900000.0   31000.0         80        188   
...        ...        ...         ...       ...        ...        ...   
19234       58         66    450000.0     600.0         74        180   
19235       55         65    275000.0     550.0         72        182   
19236       55         65    275000.0     500.0         75        186   
19237       55         67    300000.0     500.0         76        188   
19238       55         66    275000.0     500.0         74        186   

       club_team_id  club_jersey_number  preferred_foot  weak_foot  \
8               1.0                 1.0               1          3   
9               1.0                32.0               1          3   
10              1.0                33.0               1          2   
11              2.0                 1.0               1          4   
12              2.0                12.0               0          2   
...             ...                 ...             ...        ...   
19234      115820.0                23.0               0          3   
19235      115820.0                18.0               1          2   
19236      115820.0                28.0               1          2   
19237      115820.0                21.0               1          2   
19238      115820.0                15.0               1          2   

       skill_moves  international_reputation  work_rate  body_type  \
8                1                         3          0          0   
9                1                         1          0          0   
10               1                         1          0          0   
11               1                         3          0          0   
12               1                         1          0          0   
...            ...                       ...        ...        ...   
19234            2                         1          0          0   
19235            2                         1          0          0   
19236            2                         1          0          0   
19237            2                         1          0          0   
19238            2                         1          0          0   

       release_clause_eur  pace  shooting  passing  dribbling  defending  \
8              55300000.0   NaN       NaN      NaN        NaN        NaN   
9               9900000.0   NaN       NaN      NaN        NaN        NaN   
10              1900000.0   NaN       NaN      NaN        NaN        NaN   
11             66200000.0   NaN       NaN      NaN        NaN        NaN   
12              3800000.0   NaN       NaN      NaN        NaN        NaN   
...                   ...   ...       ...      ...        ...        ...   
19234            743000.0  75.0      54.0     54.0       62.0       28.0   
19235            454000.0  72.0      49.0     55.0       54.0       29.0   
19236            454000.0  66.0      51.0     48.0       55.0       20.0   
19237            495000.0  69.0      49.0     44.0       58.0       20.0   
19238            454000.0  65.0      44.0     54.0       58.0       42.0   

       physic  attacking_crossing  attacking_finishing  \
8         NaN                   9                    9   
9         NaN                  15                   14   
10        NaN                  10                    6   
11        NaN                  22                   16   
12        NaN                  11                   11   
...       ...                 ...                  ...   
19234    49.0                  58                   59   
19235    60.0                  56                   53   
19236    62.0

In [8]:
from sklearn.impute import SimpleImputer
import numpy as np

target = data_to_model.value_eur
features = data_to_model.drop(columns=["value_eur"])
# features = features.fillna(0)
mean_imputer = SimpleImputer(missing_values=np.nan, strategy='mean')
mean_imputer.fit(features)
imputed_features = mean_imputer.transform(features)

imputed_features

array([[8.3e+01, 8.4e+01, 8.7e+04, ..., 0.0e+00, 0.0e+00, 0.0e+00],
       [7.4e+01, 7.9e+01, 3.6e+04, ..., 0.0e+00, 0.0e+00, 0.0e+00],
       [6.1e+01, 7.7e+01, 3.0e+03, ..., 0.0e+00, 0.0e+00, 0.0e+00],
       ...,
       [5.5e+01, 6.5e+01, 5.0e+02, ..., 0.0e+00, 0.0e+00, 1.0e+00],
       [5.5e+01, 6.7e+01, 5.0e+02, ..., 0.0e+00, 0.0e+00, 1.0e+00],
       [5.5e+01, 6.6e+01, 5.0e+02, ..., 0.0e+00, 0.0e+00, 0.0e+00]])

In [9]:
# Normalizando os dados
# from sklearn.preprocessing import StandardScaler
#
# standerd_scaler =  StandardScaler()
# standardized_features = standerd_scaler.fit_transform(imputed_features)
# standardized_features

In [10]:
# train test split
from sklearn.model_selection import train_test_split

features_train, features_test, target_train, target_test = train_test_split(imputed_features, target, test_size=0.2)

## Comparando modelos diferentes

In [11]:
from sklearn.dummy import DummyRegressor

dummy = DummyRegressor()
dummy_model = dummy.fit(features_train, target_train)
dummy_score = dummy_model.score(features_test, target_test)

In [12]:
from sklearn.linear_model import LinearRegression

linear_regression = LinearRegression()
linear_regression_model = linear_regression.fit(features_train, target_train)
linear_regression_score = linear_regression_model.score(features_test, target_test)

In [13]:
from sklearn.ensemble import RandomForestRegressor

random_forest = RandomForestRegressor(random_state=0, n_jobs=-1, oob_score=True)
random_forest_model = random_forest.fit(features_train, target_train)
random_forest_score = random_forest_model.score(features_test, target_test)

In [14]:
score_valuation_text = f"""Score de cada modelo:
    Média {dummy_score:.4f}
    Regressão Linear {linear_regression_score:.4f}
    Floresta Aleatória {random_forest_score:.4f}
"""
print(score_valuation_text)

Score de cada modelo:
    Média -0.0007
    Regressão Linear 0.9862
    Floresta Aleatória 0.9730



### Preenchendo os dados a partir do modelo criado

In [15]:
data_to_fill = players_with_club_and_no_value.drop(columns=["sofifa_id", "player_url", "short_name", "long_name", "dob", "club_position", "value_eur"])

data_to_fill = mean_imputer.transform(data_to_fill)
predicted_values = random_forest_model.predict(data_to_fill)
predicted_values

array([ 858750.,  828500.,  890250., 3823000., 2322250.,  231150.,
        785750., 3543000.,  841000.,  216550.,  239000.,  498000.,
        379250.])

In [16]:
players_with_club_and_no_value.value_eur = predicted_values
players_with_club_and_no_value

sofifa_id                                         player_url  \
11929     115909  https://sofifa.com/player/115909/ruben-castro-...   
11987     124344  https://sofifa.com/player/124344/ibrahim-oztur...   
12264     153066  https://sofifa.com/player/153066/lucas-licht/2...   
13149     152912  https://sofifa.com/player/152912/jose-sand/220002   
13177     110381  https://sofifa.com/player/110381/maximiliano-r...   
14103     262371  https://sofifa.com/player/262371/craig-dundas/...   
14611     140222  https://sofifa.com/player/140222/paulo-da-silv...   
15125     138703  https://sofifa.com/player/138703/andres-daless...   
16065     158293  https://sofifa.com/player/158293/zhi-zheng/220002   
16498       2702  https://sofifa.com/player/2702/kevin-ellison/2...   
16814      29118  https://sofifa.com/player/29118/shinji-ono/220002   
18159      29552  https://sofifa.com/player/29552/shunsuke-nakam...   
18166     254704  https://sofifa.com/player/254704/kazuyoshi-miu...   

            short_name                     long_name  overall  potential  \
11929     Rubén Castro           Rubén Castro Martín       69         69   
11987        İ. Öztürk                İbrahim Öztürk       67         67   
12264         L. Licht            Lucas Matías Licht       69         69   
13149          J. Sand             José Gustavo Sand       75         75   
13177     M. Rodríguez   Maximiliano Rubén Rodríguez       73         73   
14103        C. Dundas                  Craig Dundas       56         56   
14611      P. Da Silva  Paulo César Da Silva Barrios       67         67   
15125  A. D'Alessandro   Andrés Nicolás D'Alessandro       75         75   
16065        Zheng Zhi                            郑智       67         67   
16498       K. Ellison           Kevin Keith Ellison       58         58   
16814           S. Ono                         小野 伸二       58         58   
18159      S. Nakamura                         中村 俊輔       63         63   
18166         K. Miura                         三浦 知良       59         59   

       value_eur  wage_eur         dob  weight_kg  height_cm  club_team_id  \
11929   858750.0    3000.0  1981-06-27         68        169      100851.0   
11987   828500.0    3000.0  1981-06-21         80        186      101006.0   
12264   890250.0    5000.0  1981-04-06         72        174      101084.0   
13149  3823000.0   13000.0  1980-07-17         79        182      110395.0   
13177  2322250.0   10000.0  1981-01-02         74        173      110396.0   
14103   231150.0    1000.0  1981-02-16         89        188      110799.0   
14611   785750.0     500.0  1980-02-01         76        180      111007.0   
15125  3543000.0     900.0  1981-04-15         69        174      111325.0   
16065   841000.0    5000.0  1980-08-20         75        180      111839.0   
16498   216550.0    2000.0  1979-02-23         76        185      112254.0   
16814   239000.0     800.0  1979-09-27         76        175      112444.0   
18159   498000.0     800.0  1978-06-24         71        178      113197.0   
18166   379250.0     700.0  1967-02-26         72        177      113197.0   

      club_position  club_jersey_number  preferred_foot  weak_foot  \
11929            ST                 7.0               1          3   
11987            CB                38.0               1          2   
12264           SUB                25.0               0          4   
13149            LS                 9.0               1          4   
13177           SUB                11.0               1          4   
14103           RES                14.0               1          3   
14611           LCB                 3.0               1          3   
15125           SUB                10.0               0          3   
16065           LCM                10.0               1          4   
16498           SUB                22.0               0          3   
16814           SUB                44.0               1          4   
18159         

In [17]:
data.loc[players_with_club_and_no_value.index, "value_eur"] = players_with_club_and_no_value.value_eur
data.loc[players_with_club_and_no_value.index]

sofifa_id                                         player_url  \
11929     115909  https://sofifa.com/player/115909/ruben-castro-...   
11987     124344  https://sofifa.com/player/124344/ibrahim-oztur...   
12264     153066  https://sofifa.com/player/153066/lucas-licht/2...   
13149     152912  https://sofifa.com/player/152912/jose-sand/220002   
13177     110381  https://sofifa.com/player/110381/maximiliano-r...   
14103     262371  https://sofifa.com/player/262371/craig-dundas/...   
14611     140222  https://sofifa.com/player/140222/paulo-da-silv...   
15125     138703  https://sofifa.com/player/138703/andres-daless...   
16065     158293  https://sofifa.com/player/158293/zhi-zheng/220002   
16498       2702  https://sofifa.com/player/2702/kevin-ellison/2...   
16814      29118  https://sofifa.com/player/29118/shinji-ono/220002   
18159      29552  https://sofifa.com/player/29552/shunsuke-nakam...   
18166     254704  https://sofifa.com/player/254704/kazuyoshi-miu...   

            short_name                     long_name  overall  potential  \
11929     Rubén Castro           Rubén Castro Martín       69         69   
11987        İ. Öztürk                İbrahim Öztürk       67         67   
12264         L. Licht            Lucas Matías Licht       69         69   
13149          J. Sand             José Gustavo Sand       75         75   
13177     M. Rodríguez   Maximiliano Rubén Rodríguez       73         73   
14103        C. Dundas                  Craig Dundas       56         56   
14611      P. Da Silva  Paulo César Da Silva Barrios       67         67   
15125  A. D'Alessandro   Andrés Nicolás D'Alessandro       75         75   
16065        Zheng Zhi                            郑智       67         67   
16498       K. Ellison           Kevin Keith Ellison       58         58   
16814           S. Ono                         小野 伸二       58         58   
18159      S. Nakamura                         中村 俊輔       63         63   
18166         K. Miura                         三浦 知良       59         59   

       value_eur  wage_eur         dob  weight_kg  height_cm  club_team_id  \
11929   858750.0    3000.0  1981-06-27         68        169      100851.0   
11987   828500.0    3000.0  1981-06-21         80        186      101006.0   
12264   890250.0    5000.0  1981-04-06         72        174      101084.0   
13149  3823000.0   13000.0  1980-07-17         79        182      110395.0   
13177  2322250.0   10000.0  1981-01-02         74        173      110396.0   
14103   231150.0    1000.0  1981-02-16         89        188      110799.0   
14611   785750.0     500.0  1980-02-01         76        180      111007.0   
15125  3543000.0     900.0  1981-04-15         69        174      111325.0   
16065   841000.0    5000.0  1980-08-20         75        180      111839.0   
16498   216550.0    2000.0  1979-02-23         76        185      112254.0   
16814   239000.0     800.0  1979-09-27         76        175      112444.0   
18159   498000.0     800.0  1978-06-24         71        178      113197.0   
18166   379250.0     700.0  1967-02-26         72        177      113197.0   

      club_position  club_jersey_number  preferred_foot  weak_foot  \
11929            ST                 7.0               1          3   
11987            CB                38.0               1          2   
12264           SUB                25.0               0          4   
13149            LS                 9.0               1          4   
13177           SUB                11.0               1          4   
14103           RES                14.0               1          3   
14611           LCB                 3.0               1          3   
15125           SUB                10.0               0          3   
16065           LCM                10.0               1          4   
16498           SUB                22.0               0          3   
16814           SUB                44.0               1          4   
18159         

In [18]:
count_na(data)

value_eur                61
wage_eur                 61
club_team_id             61
club_position            61
club_jersey_number       61
release_clause_eur     1176
pace                   2132
shooting               2132
passing                2132
dribbling              2132
defending              2132
physic                 2132
goalkeeping_speed     17107
dtype: int64

In [19]:
# Removendo jogadores não ativos
data = data.query("not club_team_id.isna()").copy()

count_na(data)

release_clause_eur     1115
pace                   2124
shooting               2124
passing                2124
dribbling              2124
defending              2124
physic                 2124
goalkeeping_speed     17054
dtype: int64

In [20]:
# TODO acidionar .joblib ou .pickle do modelo

# Prevendo pace, shooting, passing, dribbling, defending, physic dos goleiros

In [21]:
gk_data = data.query("GK == 1").copy()
gk_data

sofifa_id                                         player_url  \
8        192563  https://sofifa.com/player/192563/bernd-leno/22...   
9        233934  https://sofifa.com/player/233934/aaron-ramsdal...   
10       246671  https://sofifa.com/player/246671/arthur-okonkw...   
11       202811  https://sofifa.com/player/202811/emiliano-mart...   
12       192317  https://sofifa.com/player/192317/jed-steer/220002   
...         ...                                                ...   
2127     257618  https://sofifa.com/player/257618/armin-gremsl/...   
2128     262424  https://sofifa.com/player/262424/sorin-mogosan...   
2129     262420  https://sofifa.com/player/262420/robert-popa/2...   
2130     210022  https://sofifa.com/player/210022/philip-sprint...   
2131     244417  https://sofifa.com/player/244417/julian-krahl/...   

       short_name                 long_name  overall  potential   value_eur  \
8         B. Leno                Bernd Leno       83         84  28000000.0   
9     A. Ramsdale            Aaron Ramsdale       74         79   5000000.0   
10     A. Okonkwo            Arthur Okonkwo       61         77    750000.0   
11    E. Martínez  Damián Emiliano Martínez       84         85  33500000.0   
12       J. Steer                 Jed Steer       72         73   1900000.0   
...           ...                       ...      ...        ...         ...   
2127    A. Gremsl              Armin Gremsl       63         66    475000.0   
2128  S. Mogoșanu            Sorin Mogoșanu       58         60    170000.0   
2129      R. Popa               Robert Popa       56         75    325000.0   
2130    P. Sprint             Philip Sprint       62         64    350000.0   
2131     J. Krahl              Julian Krahl       62         76    850000.0   

      wage_eur         dob  weight_kg  height_cm  club_team_id club_position  \
8      87000.0  1992-03-04         83        190           1.0            GK   
9      36000.0  1998-05-14         77        188           1.0           SUB   
10      3000.0  2001-09-09         83        196           1.0           RES   
11     81000.0  1992-09-02         88        195           2.0            GK   
12     31000.0  1992-09-23         80        188           2.0           SUB   
...        ...         ...        ...        ...           ...           ...   
2127    2000.0  1994-08-13         85        190      115716.0           RES   
2128    1000.0  1992-11-22         77        180      115716.0            GK   
2129     500.0  2003-03-05         83        190      115716.0           SUB   
2130     700.0  1993-06-27         98        196      115820.0            GK   
2131     500.0  2000-01-22         90        194      115820.0           SUB   

      club_jersey_number  preferred_foot  weak_foot  skill_moves  \
8                    1.0               1          3            1   
9                   32.0               1          3            1   
10                  33.0               1          2            1   
11                   1.0               1          4            1   
12                  12.0               0          2            1   
...                  ...             ...        ...          ...   
2127                94.0               0          2            1   
2128                 1.0               1          1            1   
2129                31.0               1          2            1   
2130                 1.0               1          2            1   
2131                13.0               1          3            1   

      international_reputation  work_rate  body_type  release_clause_eur  \
8                            3          0          0          55300000.0   
9                            1          0          0           9900000.0   
10                           1          0          0           1900000.0   
11                           3          0          0          66200000.0   
12                           1          0          0      

In [22]:
data_without_gk = data.query("GK == 0").copy()
data_without_gk

sofifa_id                                         player_url  \
2185      188567  https://sofifa.com/player/188567/pierre-emeric...   
2186      209989  https://sofifa.com/player/209989/thomas-partey...   
2187      193301  https://sofifa.com/player/193301/alexandre-lac...   
2188      222665  https://sofifa.com/player/222665/martin-odegaa...   
2189      226110  https://sofifa.com/player/226110/nicolas-pepe/...   
...          ...                                                ...   
19234     262593  https://sofifa.com/player/262593/moritz-seiffe...   
19235     262614  https://sofifa.com/player/262614/till-muschkow...   
19236     262755  https://sofifa.com/player/262755/shalva-ogbaid...   
19237     262756  https://sofifa.com/player/262756/matteo-gumane...   
19238     263236  https://sofifa.com/player/263236/yazid-heimur/...   

           short_name                                    long_name  overall  \
2185    P. Aubameyang  Pierre-Emerick Emiliano François Aubameyang       85   
2186        T. Partey                                Thomas Partey       83   
2187     A. Lacazette                          Alexandre Lacazette       82   
2188      M. Ødegaard                              Martin Ødegaard       82   
2189          N. Pépé                                 Nicolas Pépé       81   
...               ...                                          ...      ...   
19234     M. Seiffert                              Moritz Seiffert       58   
19235  T. Muschkowski                             Till Muschkowski       55   
19236     S. Ogbaidze                              Shalva Ogbaidze       55   
19237      M. Gumaneh                               Matteo Gumaneh       55   
19238       Y. Heimur                                 Yazid Heimur       55   

       potential   value_eur  wage_eur         dob  weight_kg  height_cm  \
2185          85  43500000.0  145000.0  1989-06-18         80        187   
2186          83  36500000.0  115000.0  1993-06-13         77        185   
2187          82  30000000.0  110000.0  1991-05-28         73        175   
2188          88  49000000.0   89000.0  1998-12-17         68        178   
2189          82  30500000.0   95000.0  1995-05-29         73        183   
...          ...         ...       ...         ...        ...        ...   
19234         66    450000.0     600.0  2000-11-04         74        180   
19235         65    275000.0     550.0  2001-05-06         72        182   
19236         65    275000.0     500.0  2002-08-01         75        186   
19237         67    300000.0     500.0  2002-09-05         76        188   
19238         66    275000.0     500.0  2002-11-18         74        186   

       club_team_id club_position  club_jersey_number  preferred_foot  \
2185            1.0            ST                14.0               1   
2186            1.0           RDM                 5.0               1   
2187            1.0           SUB                 9.0               1   
2188            1.0           CAM                 8.0               0   
2189            1.0            RM                19.0               0   
...             ...           ...                 ...             ...   
19234      115820.0           SUB                23.0               0   
19235      115820.0           SUB                18.0               1   
19236      115820.0           RES                28.0               1   
19237      115820.0           SUB                21.0               1   
19238      115820.0           RES                15.0               1   

       weak_foot  skill_moves  international_reputation  work_rate  body_type  \
2185           4            4                         3          0          0   
2186           3            3                         2          0          0   
2187           4            4                         3          1          0   
2188           2            5                         3          0          0   
2189      

In [23]:
columns_to_predict = ["pace", "shooting", "passing", "dribbling", "defending", "physic"]


data_to_model = data_without_gk.drop(columns=["sofifa_id", "player_url", "short_name", "long_name", "dob", "club_position"])

data_to_predict = gk_data.drop(columns=["sofifa_id", "player_url", "short_name", "long_name", "dob", "club_position", *columns_to_predict])

for column_to_predict in columns_to_predict:

    target = data_to_model[column_to_predict]
    features = data_to_model.drop(columns=columns_to_predict)
    mean_imputer = SimpleImputer(missing_values=np.nan, strategy='mean')
    mean_imputer.fit(features)
    imputed_features = mean_imputer.transform(features)

    features_train, features_test, target_train, target_test = train_test_split(imputed_features, target, test_size=0.2)

    linear_regression = LinearRegression()
    linear_regression_model = linear_regression.fit(features_train, target_train)
    linear_regression_score = linear_regression_model.score(features_test, target_test)

    predicted_values = linear_regression_model.predict(mean_imputer.transform(data_to_predict))
    gk_data[column_to_predict] = predicted_values.round()
    data.loc[gk_data.index, column_to_predict] = gk_data[column_to_predict]

    print(f"{column_to_predict}: score: {linear_regression_score:.4f}")

data.loc[gk_data.index]

pace: score: 0.9993
shooting: score: 0.9991
passing: score: 0.9961
dribbling: score: 0.9991
defending: score: 0.9997
physic: score: 0.9991


sofifa_id                                         player_url  \
8        192563  https://sofifa.com/player/192563/bernd-leno/22...   
9        233934  https://sofifa.com/player/233934/aaron-ramsdal...   
10       246671  https://sofifa.com/player/246671/arthur-okonkw...   
11       202811  https://sofifa.com/player/202811/emiliano-mart...   
12       192317  https://sofifa.com/player/192317/jed-steer/220002   
...         ...                                                ...   
2127     257618  https://sofifa.com/player/257618/armin-gremsl/...   
2128     262424  https://sofifa.com/player/262424/sorin-mogosan...   
2129     262420  https://sofifa.com/player/262420/robert-popa/2...   
2130     210022  https://sofifa.com/player/210022/philip-sprint...   
2131     244417  https://sofifa.com/player/244417/julian-krahl/...   

       short_name                 long_name  overall  potential   value_eur  \
8         B. Leno                Bernd Leno       83         84  28000000.0   
9     A. Ramsdale            Aaron Ramsdale       74         79   5000000.0   
10     A. Okonkwo            Arthur Okonkwo       61         77    750000.0   
11    E. Martínez  Damián Emiliano Martínez       84         85  33500000.0   
12       J. Steer                 Jed Steer       72         73   1900000.0   
...           ...                       ...      ...        ...         ...   
2127    A. Gremsl              Armin Gremsl       63         66    475000.0   
2128  S. Mogoșanu            Sorin Mogoșanu       58         60    170000.0   
2129      R. Popa               Robert Popa       56         75    325000.0   
2130    P. Sprint             Philip Sprint       62         64    350000.0   
2131     J. Krahl              Julian Krahl       62         76    850000.0   

      wage_eur         dob  weight_kg  height_cm  club_team_id club_position  \
8      87000.0  1992-03-04         83        190           1.0            GK   
9      36000.0  1998-05-14         77        188           1.0           SUB   
10      3000.0  2001-09-09         83        196           1.0           RES   
11     81000.0  1992-09-02         88        195           2.0            GK   
12     31000.0  1992-09-23         80        188           2.0           SUB   
...        ...         ...        ...        ...           ...           ...   
2127    2000.0  1994-08-13         85        190      115716.0           RES   
2128    1000.0  1992-11-22         77        180      115716.0            GK   
2129     500.0  2003-03-05         83        190      115716.0           SUB   
2130     700.0  1993-06-27         98        196      115820.0            GK   
2131     500.0  2000-01-22         90        194      115820.0           SUB   

      club_jersey_number  preferred_foot  weak_foot  skill_moves  \
8                    1.0               1          3            1   
9                   32.0               1          3            1   
10                  33.0               1          2            1   
11                   1.0               1          4            1   
12                  12.0               0          2            1   
...                  ...             ...        ...          ...   
2127                94.0               0          2            1   
2128                 1.0               1          1            1   
2129                31.0               1          2            1   
2130                 1.0               1          2            1   
2131                13.0               1          3            1   

      international_reputation  work_rate  body_type  release_clause_eur  \
8                            3          0          0          55300000.0   
9                            1          0          0           9900000.0   
10                           1          0          0           1900000.0   
11                           3          0          0          66200000.0   
12                           1          0          0      

In [24]:
count_na(data)

release_clause_eur     1115
goalkeeping_speed     17054
dtype: int64

## Prevendo goalkeeping_speed dos jogadores

In [25]:
players_data = data.query("GK == 0").copy()
players_data

sofifa_id                                         player_url  \
2185      188567  https://sofifa.com/player/188567/pierre-emeric...   
2186      209989  https://sofifa.com/player/209989/thomas-partey...   
2187      193301  https://sofifa.com/player/193301/alexandre-lac...   
2188      222665  https://sofifa.com/player/222665/martin-odegaa...   
2189      226110  https://sofifa.com/player/226110/nicolas-pepe/...   
...          ...                                                ...   
19234     262593  https://sofifa.com/player/262593/moritz-seiffe...   
19235     262614  https://sofifa.com/player/262614/till-muschkow...   
19236     262755  https://sofifa.com/player/262755/shalva-ogbaid...   
19237     262756  https://sofifa.com/player/262756/matteo-gumane...   
19238     263236  https://sofifa.com/player/263236/yazid-heimur/...   

           short_name                                    long_name  overall  \
2185    P. Aubameyang  Pierre-Emerick Emiliano François Aubameyang       85   
2186        T. Partey                                Thomas Partey       83   
2187     A. Lacazette                          Alexandre Lacazette       82   
2188      M. Ødegaard                              Martin Ødegaard       82   
2189          N. Pépé                                 Nicolas Pépé       81   
...               ...                                          ...      ...   
19234     M. Seiffert                              Moritz Seiffert       58   
19235  T. Muschkowski                             Till Muschkowski       55   
19236     S. Ogbaidze                              Shalva Ogbaidze       55   
19237      M. Gumaneh                               Matteo Gumaneh       55   
19238       Y. Heimur                                 Yazid Heimur       55   

       potential   value_eur  wage_eur         dob  weight_kg  height_cm  \
2185          85  43500000.0  145000.0  1989-06-18         80        187   
2186          83  36500000.0  115000.0  1993-06-13         77        185   
2187          82  30000000.0  110000.0  1991-05-28         73        175   
2188          88  49000000.0   89000.0  1998-12-17         68        178   
2189          82  30500000.0   95000.0  1995-05-29         73        183   
...          ...         ...       ...         ...        ...        ...   
19234         66    450000.0     600.0  2000-11-04         74        180   
19235         65    275000.0     550.0  2001-05-06         72        182   
19236         65    275000.0     500.0  2002-08-01         75        186   
19237         67    300000.0     500.0  2002-09-05         76        188   
19238         66    275000.0     500.0  2002-11-18         74        186   

       club_team_id club_position  club_jersey_number  preferred_foot  \
2185            1.0            ST                14.0               1   
2186            1.0           RDM                 5.0               1   
2187            1.0           SUB                 9.0               1   
2188            1.0           CAM                 8.0               0   
2189            1.0            RM                19.0               0   
...             ...           ...                 ...             ...   
19234      115820.0           SUB                23.0               0   
19235      115820.0           SUB                18.0               1   
19236      115820.0           RES                28.0               1   
19237      115820.0           SUB                21.0               1   
19238      115820.0           RES                15.0               1   

       weak_foot  skill_moves  international_reputation  work_rate  body_type  \
2185           4            4                         3          0          0   
2186           3            3                         2          0          0   
2187           4            4                         3          1          0   
2188           2            5                         3          0          0   
2189      

In [26]:
data_to_predict = players_data.drop(columns=["sofifa_id", "player_url", "short_name", "long_name", "dob", "goalkeeping_speed", "club_position"])

In [27]:
features = gk_data.drop(columns=["sofifa_id", "player_url", "short_name", "long_name", "dob", "goalkeeping_speed", "club_position"])
target = gk_data.goalkeeping_speed

mean_imputer = SimpleImputer(missing_values=np.nan, strategy='mean')
mean_imputer.fit(features)
imputed_features = mean_imputer.transform(features)

features_train, features_test, target_train, target_test = train_test_split(imputed_features, target, test_size=0.2)


linear_regression = LinearRegression()
linear_regression_model = linear_regression.fit(features_train, target_train)
linear_regression_score = linear_regression_model.score(features_test, target_test)

print(f"goalkeeping_speed: score: {linear_regression_score:.4f}")

predicted_values = linear_regression_model.predict(mean_imputer.transform(data_to_predict))
players_data["goalkeeping_speed"] = predicted_values.round()
data.loc[players_data.index, "goalkeeping_speed"] = players_data["goalkeeping_speed"]

goalkeeping_speed: score: 0.9993


In [29]:
data.loc[players_data.index]

sofifa_id                                         player_url  \
2185      188567  https://sofifa.com/player/188567/pierre-emeric...   
2186      209989  https://sofifa.com/player/209989/thomas-partey...   
2187      193301  https://sofifa.com/player/193301/alexandre-lac...   
2188      222665  https://sofifa.com/player/222665/martin-odegaa...   
2189      226110  https://sofifa.com/player/226110/nicolas-pepe/...   
...          ...                                                ...   
19234     262593  https://sofifa.com/player/262593/moritz-seiffe...   
19235     262614  https://sofifa.com/player/262614/till-muschkow...   
19236     262755  https://sofifa.com/player/262755/shalva-ogbaid...   
19237     262756  https://sofifa.com/player/262756/matteo-gumane...   
19238     263236  https://sofifa.com/player/263236/yazid-heimur/...   

           short_name                                    long_name  overall  \
2185    P. Aubameyang  Pierre-Emerick Emiliano François Aubameyang       85   
2186        T. Partey                                Thomas Partey       83   
2187     A. Lacazette                          Alexandre Lacazette       82   
2188      M. Ødegaard                              Martin Ødegaard       82   
2189          N. Pépé                                 Nicolas Pépé       81   
...               ...                                          ...      ...   
19234     M. Seiffert                              Moritz Seiffert       58   
19235  T. Muschkowski                             Till Muschkowski       55   
19236     S. Ogbaidze                              Shalva Ogbaidze       55   
19237      M. Gumaneh                               Matteo Gumaneh       55   
19238       Y. Heimur                                 Yazid Heimur       55   

       potential   value_eur  wage_eur         dob  weight_kg  height_cm  \
2185          85  43500000.0  145000.0  1989-06-18         80        187   
2186          83  36500000.0  115000.0  1993-06-13         77        185   
2187          82  30000000.0  110000.0  1991-05-28         73        175   
2188          88  49000000.0   89000.0  1998-12-17         68        178   
2189          82  30500000.0   95000.0  1995-05-29         73        183   
...          ...         ...       ...         ...        ...        ...   
19234         66    450000.0     600.0  2000-11-04         74        180   
19235         65    275000.0     550.0  2001-05-06         72        182   
19236         65    275000.0     500.0  2002-08-01         75        186   
19237         67    300000.0     500.0  2002-09-05         76        188   
19238         66    275000.0     500.0  2002-11-18         74        186   

       club_team_id club_position  club_jersey_number  preferred_foot  \
2185            1.0            ST                14.0               1   
2186            1.0           RDM                 5.0               1   
2187            1.0           SUB                 9.0               1   
2188            1.0           CAM                 8.0               0   
2189            1.0            RM                19.0               0   
...             ...           ...                 ...             ...   
19234      115820.0           SUB                23.0               0   
19235      115820.0           SUB                18.0               1   
19236      115820.0           RES                28.0               1   
19237      115820.0           SUB                21.0               1   
19238      115820.0           RES                15.0               1   

       weak_foot  skill_moves  international_reputation  work_rate  body_type  \
2185           4            4                         3          0          0   
2186           3            3                         2          0          0   
2187           4            4                         3          1          0   
2188           2            5                         3          0          0   
2189      

In [28]:
count_na(data)

release_clause_eur    1115
dtype: int64